File: src/main/java/com/mestrap/action/CommandAction.java
package com.mestrap.action;

import com.mestrap.core.ActionContext;
import com.mestrap.ssh.JschCommandExecutor;
import com.mestrap.exception.TaskException;
import com.mestrap.utils.LogPrinter;
import com.mestrap.utils.VariableReplacer;
import org.apache.commons.cli.Option;

import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * 执行远程命令。
 * -e, --execute 远程执行的命令内容，支持 ${...} 格式参数
 *
 * @author melvek
 */
public class CommandAction implements TaskAction {

    @Override
    public String name() {
        return "command";
    }

    @Override
    public List<Option> cliOptions() {
        return Collections.singletonList(
                Option.builder("e").longOpt("execute").hasArg().argName("string")
                        .desc("Command string (injected as ${command})")
                        .build()
        );
    }

    @Override
    public Map<String, String> cliVarMapping() {
        return Collections.singletonMap("execute", "command");
    }

    @Override
    public void execute(ActionContext ctx) throws Exception {
        Object raw = ctx.getWith().get("command");
        if (raw == null) {
            throw new TaskException("command action requires 'command' parameter");
        }

        String cmd = VariableReplacer.replace(String.valueOf(raw), ctx.getVars());

        LogPrinter.info("Execute command: " + cmd);
        LogPrinter.emptyLine();

        int exitCode = JschCommandExecutor.executeCommand(
                ctx.getHostVars().getHost(),
                ctx.getHostVars().getPort(),
                ctx.getHostVars().getUserName(),
                ctx.getHostVars().getPassword(),
                cmd
        );

        if (exitCode != 0) {
            throw new TaskException("Command failed with exit code " + exitCode);
        }
    }
}

File: internal/action/CommandAction.go
package action

import (
	"fmt"

	"mestrap.com/taskssh/internal/ssh"
)

// CommandAction 执行远程命令。
type CommandAction struct{}

func (a *CommandAction) Name() string { return "command" }

func (a *CommandAction) Execute(ctx *Context) error {
	raw, ok := ctx.With["command"]
	if !ok {
		return fmt.Errorf("command action requires 'command' parameter")
	}

	cmd := fmt.Sprintf("%v", raw)
	cmd, err := ctx.Vars.Replace(cmd)
	if err != nil {
		return err
	}

	client, err := ssh.Connect(ctx.Host)
	if err != nil {
		return err
	}
	defer client.Close()

	result, err := client.Exec(cmd)
	if err != nil {
		return err
	}

	if result.ExitCode != 0 {
		return fmt.Errorf("command failed with exit code %d", result.ExitCode)
	}
	return nil
}
